pkg/volume/rbd/rbd_util.go

```diff
@@ -0,0 +1,145 @@
+	for i := 0; i < maxRetries; i++ {
+		_, err := os.Stat(devicePath)
+		if err == nil {
+			return true
+		}
+		if err != nil && !os.IsNotExist(err) {
+			return false
+		}
+		time.Sleep(time.Second)
+	}
+	return false
+}
+
+// make a directory like /var/lib/kubelet/plugins/kubernetes.io/pod/rbd/pool-image-image
+func makePDNameInternal(host volume.VolumeHost, pool string, image string) string {
+	return path.Join(host.GetPluginDir(RBDPluginName), "rbd", pool+"-image-"+image)
+}
+
+type RBDUtil struct{}
+
+func (util *RBDUtil) MakeGlobalPDName(rbd rbd) string {
+	return makePDNameInternal(rbd.plugin.host, rbd.pool, rbd.image)
+}
+
+func (util *RBDUtil) AttachDisk(rbd rbd) error {
+	var err error
+	devicePath := strings.Join([]string{"/dev/rbd", rbd.pool, rbd.image}, "/")
+	exist := waitForPathToExist(devicePath, 1)
+	if exist == false {
+		// modprobe
+		_, err = rbd.plugin.execCommand("modprobe", []string{"rbd"})
+		if err != nil {
+			glog.Errorf("rbd: failed to modprobe rbd error:%v", err)
+			return err
+		}
+		// rbd map
+		l := len(rbd.mon)
+		// avoid mount storm, pick a host randomly
+		start := rand.Int() % l
+		// iterate all hosts until mount succeeds.
+		for i := start; i < start+l; i++ {
+			mon := rbd.mon[i%l]
+			glog.V(1).Infof("rbd: map mon %s", mon)
+			if rbd.secret != "" {
+				_, err = rbd.plugin.execCommand("rbd",
+					[]string{"map", rbd.image, "--pool", rbd.pool, "--id", rbd.id, "-m", mon, "--key=" + rbd.secret})
+			} else {
+				_, err = rbd.plugin.execCommand("rbd",
+					[]string{"map", rbd.image, "--pool", rbd.pool, "--id", rbd.id, "-m", mon, "-k", rbd.keyring})
+			}
+			if err == nil {
+				break
+			}
+		}
+	}
+	if err != nil {
+		glog.Errorf("rbd: failed to map image:Error: %v", err)
+		return err
+	}
+	exist = waitForPathToExist(devicePath, 10)
+	if !exist {
+		return errors.New("Could not map image: Timeout after 10s")
+	}
+	// mount it
+	globalPDPath := rbd.manager.MakeGlobalPDName(rbd)
+	mountpoint, err := rbd.mounter.IsMountPoint(globalPDPath)
+	if mountpoint {
+		glog.Infof("rbd: %s already mounted", globalPDPath)
+		return nil
+	}
+
+	if err := os.MkdirAll(globalPDPath, 0750); err != nil {
+		glog.Errorf("rbd: failed to mkdir %s, error", globalPDPath)
+		return err
+	}
+
+	err = rbd.mounter.Mount(devicePath, globalPDPath, rbd.fsType, nil)
+	if err != nil {
+		glog.Errorf("rbd: failed to mount rbd volume %s [%s] to %s, error %v", devicePath, rbd.fsType, globalPDPath, err)
+	}
+
+	return err
+}
+
+func (util *RBDUtil) DetachDisk(rbd rbd, mntPath string) error {
+	device, cnt, err := mount.GetDeviceNameFromMount(rbd.mounter, mntPath)
+	if err != nil {
+		glog.Errorf("rbd detach disk: failed to get device from mnt: %s\nError: %v", mntPath, err)
+		return err
+	}
+	if err = rbd.mounter.Unmount(mntPath); err != nil {
+		glog.Errorf("rbd detach disk: failed to umount: %s\nError: %v", mntPath, err)
+		return err
+	}
+	cnt--
+	// if device is no longer used, see if can unmap
+	if cnt == 0 {
```

Comment:
It would be more idiomatic go to have "if cnt <= 1" here than a decrement then comparison to zero.